File: app/src/main/java/com/erotc/learning/parser/QuestionParser.java
package com.erotc.learning.parser;

import com.erotc.learning.dao.QuestionEntry;
import com.erotc.learning.dao.QuestionSet;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;

/**
 * Created on 11/26/2018.
 */
public class QuestionParser {
    private static final String FIELD_LABEL = "label";
    private static final String FIELD_LEVEL = "level";
    private static final String FIELD_LOCKED = "locked";
    private static final String FIELD_POINTS_TO_PROCEED = "pointsToProceed";
    private static final String FIELD_ANSWER_SET = "answerSet";
    private static final String FIELD_QUESTION_ENTRIES = "questionEntries";
    private static final String FIELD_QUESTION = "question";
    private static final String FIELD_ANSWER = "answer";

    private InputStream inputStream;
    private ByteArrayOutputStream byteArrayOutputStream;

    public QuestionParser(InputStream inputStream) throws IOException {
        this.inputStream = inputStream;

        loadDataToByteArrayOutputStream();
    }

    private void loadDataToByteArrayOutputStream() throws IOException {
        byteArrayOutputStream = new ByteArrayOutputStream();

        int dataTye = inputStream.read();

        try {
            while (dataTye != -1) {
                byteArrayOutputStream.write(dataTye);
                dataTye = inputStream.read();
            }
        } finally {
            inputStream.close();
        }
    }

    public ArrayList<QuestionSet> parse() throws JSONException {
        ArrayList<QuestionSet> questionSets = new ArrayList<>();

        JSONArray jArray = new JSONArray(byteArrayOutputStream.toString());

        for (int i = 0; i < jArray.length(); i++) {
            JSONObject jObject = jArray.getJSONObject(i);
            questionSets.add(readQuestionSet(jObject));
        }

        return questionSets;
    }

    private QuestionSet readQuestionSet(JSONObject jObject) throws JSONException {
        QuestionSet questionSet = new QuestionSet();
        questionSet.setLabel(jObject.getString(FIELD_LABEL));
        questionSet.setLevel(jObject.getInt(FIELD_LEVEL));
        questionSet.setLocked(jObject.getBoolean(FIELD_LOCKED));
        questionSet.setPointsToProceed(jObject.getInt(FIELD_POINTS_TO_PROCEED));
        questionSet.setAnswerSet(jObject.getString(FIELD_ANSWER_SET));
        questionSet.setQuestionArray(readQuestionEntryArray(jObject.getJSONArray(FIELD_QUESTION_ENTRIES)));

        return questionSet;
    }

    private ArrayList<QuestionEntry> readQuestionEntryArray(JSONArray jArray) throws JSONException {
        ArrayList<QuestionEntry> questionEntries = new ArrayList<>();

        for (int i = 0; i < jArray.length(); i++) {
            JSONObject jObject = jArray.getJSONObject(i);
            questionEntries.add(readQuestionEntry(jObject));
        }

        return questionEntries;
    }

    private QuestionEntry readQuestionEntry(JSONObject jObject) throws JSONException {
        QuestionEntry questionEntry = new QuestionEntry();
        questionEntry.setQuestion(jObject.getString(FIELD_QUESTION));
        questionEntry.setAnswer(jObject.getString(FIELD_ANSWER));

        return questionEntry;
    }
}

File: app/src/main/java/com/erotc/learning/parser/QuestionParser.kt
package com.erotc.learning.parser

import com.erotc.learning.data.QuestionEntry
import com.erotc.learning.data.QuestionSet
import com.erotc.learning.data.QuestionSetEntryList
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.util.*

/**
 * Created on 11/26/2018.
 */
class QuestionParser(private val inputStream: InputStream) {
    private var byteArrayOutputStream: ByteArrayOutputStream? = null

    private fun loadDataToByteArrayOutputStream() {
        byteArrayOutputStream = ByteArrayOutputStream()
        var dataTye = inputStream.read()

        inputStream.use { inputStream ->
            while (dataTye != -1) {
                byteArrayOutputStream?.write(dataTye)
                dataTye = inputStream.read()
            }
        }
    }

    fun parse(): ArrayList<QuestionSetEntryList> {
        val questionSets = ArrayList<QuestionSetEntryList>()

        val jArray = JSONArray(byteArrayOutputStream.toString())
        for (i in 0 until jArray.length()) {
            val jObject = jArray.getJSONObject(i)
            questionSets.add(readQuestionSet(jObject))
        }
        return questionSets
    }

    private fun readQuestionSet(jObject: JSONObject): QuestionSetEntryList {
        val questionSet = QuestionSet()
        questionSet.label = jObject.getString(FIELD_LABEL)
        questionSet.level = jObject.getInt(FIELD_LEVEL)
        questionSet.isLocked = jObject.getBoolean(FIELD_LOCKED)
        questionSet.pointsToProceed = jObject.getInt(FIELD_POINTS_TO_PROCEED)
        questionSet.answerSet = jObject.getString(FIELD_ANSWER_SET)
        val questionEntries = readQuestionEntryArray(jObject.getJSONArray(FIELD_QUESTION_ENTRIES))

        return QuestionSetEntryList(questionSet, questionEntries)
    }

    private fun readQuestionEntryArray(jArray: JSONArray): ArrayList<QuestionEntry> {
        val questionEntries = ArrayList<QuestionEntry>()
        for (i in 0 until jArray.length()) {
            val jObject = jArray.getJSONObject(i)
            questionEntries.add(readQuestionEntry(jObject))
        }
        return questionEntries
    }

    private fun readQuestionEntry(jObject: JSONObject): QuestionEntry {
        val questionEntry = QuestionEntry()
        questionEntry.question = jObject.getString(FIELD_QUESTION)
        questionEntry.answer = jObject.getString(FIELD_ANSWER)
        return questionEntry
    }

    companion object {
        private const val FIELD_LABEL = "label"
        private const val FIELD_LEVEL = "level"
        private const val FIELD_LOCKED = "locked"
        private const val FIELD_POINTS_TO_PROCEED = "pointsToProceed"
        private const val FIELD_ANSWER_SET = "answerSet"
        private const val FIELD_QUESTION_ENTRIES = "questionEntries"
        private const val FIELD_QUESTION = "question"
        private const val FIELD_ANSWER = "answer"
    }

    init {
        loadDataToByteArrayOutputStream()
    }
}